_posts/2025-03-04-fast-streaming-inserts-in-duckdb-with-adbc.md

```diff
@@ -0,0 +1,201 @@
+
+The first two either ran into issues of data locality or weren't optimized enough:
+
+*    **CustomArrows** : functions to run on each Arrow Record to create a new Record to insert along with the original
+*    **DuckRunner** : run a series of queries on the database file before rotation
+
+
+Reasoning that if unnesting deeply nested data in Arrow Record arrays was causing data locality issues:
+
+*   **Normalizer**: a Bufarrow API used in the in the deserialization function to normalize the message data and append it to another Arrow Record, inserted into a separate table
+
+This approach allowed throughput to go back to levels almost as high as without Normalizer \- flat data is much faster to process and insert.
+
+# Oh, we're halfway there...livin' on a prayer
+
+Next, I tried opening concurrent connections to multiple databases. **BAM\!** ***Segfault***. DuckDB concurrency model isn't [designed](https://duckdb.org/docs/stable/connect/concurrency.html#handling-concurrency) that way. From within a process only a single database (in-memory or file) can be opened, then other database files can be [attached](https://duckdb.org/docs/stable/sql/statements/attach.html) to the central db's catalog. 
+
+Having already decided to rotate DB files, I decided to make a separate program ([Runner](https://github.com/loicalleyne/quacfka-runner)) to process the database files as they were rotated, running aggregations on normalized data and table dumps to parquet. This meant setting up an RPC connection between the two and figuring out a backpressure mechanism to avoid `disk full` events.
+
+However having the two running simultaneously was causing memory pressure issues, not to mention massively slowing down the throughput. Upgrading the VM to one with more vCPUs and memory only helped a little, there was clearly some resource contention going on.
+
+Since Go 1.5, the default GOMAXPROCS value is the number of CPU cores available. What if this was reduced to "sandbox" the ingestion process, along with setting the DuckDB thread count in the Runner? This actually worked so well, it increased the overall throughput. [Runner](https://github.com/loicalleyne/quacfka-runner) runs the `COPY...TO...parquet` queries, walks the parquet output folder, uploads files to object storage and deletes the uploaded files. Balancing the DuckDB file rotation size threshold in [Quafka-Service](https://github.com/loicalleyne/quacfka-service) allows Runner to keep up and avoid a backlog of DB files on disk. 
+
+# Results
+
+<img src="{{ site.baseurl }}/img/adbc-duckdb/btop.png" width="100%" class="img-responsive" alt="btop utility showing CPU and memory usage of quacfka-service and runner" aria-hidden="true">  
+Note: both runs with GOMAXPROCS set to 24 (the number of DuckDB insertion routines)
+
+Ingesting the raw data (14 fields with one deeply nested LIST.STRUCT.LIST field) \+ normalized data:  
```

Comment:
Can you make this and the section below monospaced? (Using HTML to allow you to preserve the bold text, as described above)